<h1><center>Laboratorio 3: El Pandas no muerde 🐼</center></h1>

<center><strong>MDS7202: Laboratorio de Programación Científica para Ciencia de Datos</strong></center>

### Cuerpo Docente:

- Profesores: Ignacio Meza, Sebastián Tinoco
- Auxiliar: Catherine Benavides
- Ayudante: Nicolás Ojeda, Eduardo Moya

### Equipo: SUPER IMPORTANTE - notebooks sin nombre no serán revisados

- Nombre de alumno 1: Manuel Zamorano
- Nombre de alumno 2: Javier Urrutia


### **Link de repositorio de GitHub:** `https://github.com/JUrrutia10/Laboratorios-MDS`

### Reglas:

- **Grupos de 2 personas**
- Fecha de entrega: 7 días desde la publicación, 3 días de atraso con 1 punto de descuento c/u.
- Instrucciones del lab el viernes a las 16:15 en formato online. Asistencia no es obligatoria.
- <u>Prohibidas las copias</u>. Cualquier intento de copia será debidamente penalizado con el reglamento de la escuela.
- Tienen que subir el laboratorio a u-cursos Y a su repositorio de github. Labs que no estén en u-cursos no serán revisados. Recuerden que el repositorio también tiene nota.
- Cualquier duda fuera del horario de clases al foro. Mensajes al equipo docente serán respondidos por este medio.
- Pueden usar cualquer material del curso que estimen conveniente.

### Temas a tratar

- Introducción al manejo de datos tabulares por medio de la libreria `pandas`.

### Objetivos principales del laboratorio

- Entender, aplicar y aprovechar las ventajas que nos ofrece la libreria `pandas` para manejar datos tabulares.
- Aprender a utilizar la documentación de pandas (y en general, de cualquier librería de python) como también a buscar soluciones en la web.

> **Nota**: El laboratorio deberá ser desarrollado sin el uso indiscriminado de iteradores nativos de python (aka "for", "while"). La idea es que aprendan a exprimir al máximo las funciones optimizadas que nos entrega `pandas`, las cuales vale mencionar, son bastante más eficientes que los iteradores nativos sobre DataFrames.

### Librerias utiles para el Lab 😸

In [1]:
# Libreria Core del lab.
import numpy as np
import pandas as pd
from IPython.display import display

#Libreria para plotear
import plotly.express as px

# Librerias utiles
from sklearn.preprocessing import StandardScaler

# 1. Rendimiento en Estudiantes 📚

Para este laboratorio deberán trabajar con el dataset `students_grades.json` el cual contiene información sobre el rendimiento de alumnos en distintos ramos.
El objetivo principal será el de realizar un trabajo de limpieza de prepración y exploración de datos utilizando exclusivamente `DataFrames` y sus métodos asociados.




In [2]:
# Si usted está utilizando Colabolatory le puede ser útil este código para cargar los archivos.
try:
    from google.colab import drive
    drive.mount("/content/drive", force_remount=True)
    path = '/content/drive/MyDrive/Lab programación cientifica/lab3'
except:
    print('Ignorando conexión drive-colab')

Mounted at /content/drive


## 1. Carga e Inspección de Datos [1.0 Puntos]


<p align="center">
  <img src="https://drive.google.com/uc?export=view&id=1hZZaPQlZe4VupFOXt76_wjuUA5NpUhrw" width="300" height="300">
</p>

*Don Caguayo*, rector de la Universidad de la Cachaña, les comenta que actualmente la Universidad no tiene una subdirección de análisis de datos, pero que tiene un gran interés en formar una. Es por esto, que se contacta con ud. y su equipo para iniciar de forma exploratoria diversas tareas de análisis y consolidación de datos respectivos al rendimiento de sus alumnos.

Previo a su primera reunión con el rector, el rector le envía una copia de la base de datos, la cuál a usted y su equipo le encantaría examinar. Para esto, la primera tarea que debe realizar es abrir y cargar el dataset.


> **Notar que es sumamente importante el formato que posee cada columna** para su manipulación y que posiblemente ustedes deberán especificar algunos tipos de datos a mano. Ver la documentación del método `read_json` para mayor información.

A continuación se muestra un ejemplo de los dos primeros registros del dataset:

```json
[
  {
    "names": "Rita Courtney",
    "gender": "female",
    "race/ethnicity": "group B",
    "parental level of education": "some high school",
    "lunch": "standard",
    "test preparation course": "none",
    "math score": 37,
    "reading score": 46,
    "writing score": 46
  },
  {
    "names": "Madeline Fuller",
    "gender": "female",
    "race/ethnicity": "group C",
    "parental level of education": "some high school",
    "lunch": "standard",
    "test preparation course": "none",
    "math score": 124,
    "reading score": 142,
    "writing score": 142
  }
]
```





In [3]:
df_grades =pd.read_json ('/content/drive/MyDrive/Lab programación cientifica/lab3/students_grades.json')
display(df_grades)
df_grades.info()

names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
1      Madeline Fuller  female        group C            some high school   
2        Daniel Rester    male        group E          associate's degree   
3     Charles Linstrom    male        group A           bachelor's degree   
4       Geraldine Hale  female        group A          associate's degree   
...                ...     ...            ...                         ...   
1395  Alexander Dillon    male        group D          associate's degree   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1398      Joseph Mccoy    male        group D                some college   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course math score reading score  \
0         standard                    none         37            46   
1         standard                    none        124           142   
2         standard                    none        156           196   
3         standard               completed         80            78   
4         standard                    none        131           166   
...            ...                     ...        ...           ...   
1395      standard                    none        193           142   
1396  free/reduced               completed         66            74   
1397      standard               completed         54            61   
1398  free/reduced                    none        193           196   
1399      standard               completed         68            80   

     writing score  
0               46  
1              142  
2              193  
3               81  
4              166  
...            ...  
1395           193  
1396            81  
1397            62  
1398           193  
1399            76  

[1400 rows x 9 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1400 entries, 0 to 1399
Data columns (total 9 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   names                        1400 non-null   object
 1   gender                       1400 non-null   object
 2   race/ethnicity               1400 non-null   object
 3   parental level of education  1400 non-null   object
 4   lunch                        1400 non-null   object
 5   test preparation course      1400 non-null   object
 6   math score                   1400 non-null   object
 7   reading score                1400 non-null   object
 8   writing score                1400 non-null   object
dtypes: object(9)
memory usage: 98.6+ KB


In [4]:
# Convertir columnas numéricas a tipos de datos numéricos
numeric_cols = ['math score', 'reading score', 'writing score']
df_grades[numeric_cols] = df_grades[numeric_cols].apply(pd.to_numeric, errors='coerce')
df_grades.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1400 entries, 0 to 1399
Data columns (total 9 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   names                        1400 non-null   object 
 1   gender                       1400 non-null   object 
 2   race/ethnicity               1400 non-null   object 
 3   parental level of education  1400 non-null   object 
 4   lunch                        1400 non-null   object 
 5   test preparation course      1400 non-null   object 
 6   math score                   1360 non-null   float64
 7   reading score                1360 non-null   float64
 8   writing score                1353 non-null   float64
dtypes: float64(3), object(6)
memory usage: 98.6+ KB



<p align="center">
  <img src="https://i.giphy.com/media/4FQMuOKR6zQRO/giphy.webp">
</p>


Posteriormente, el rector les comenta que la sección de TI posee una base de datos con más de **875** filas, en donde cada fila representa un alumno matriculado. Sin embargo, cree fuertemente que existen errores en esta: solo tiene registro de **875** matriculas. Por este motivo, especula que se han realizado malas prácticas al registrar dichos datos.

En consecuencia, el rector les solicita a ustedes como tarea inicial que exploren la base de datos y encuentre posibles falencias dentro de sus registros. Para esto, usted debe estar al tanto que las notas están en una escala que van del **0 al 100** (porcentaje de logro) y **no es posible que un alumno haya dado más de una vez el mismo examen**. Cualquier alumno que posea un valor diferente a esta escala, debería ser eliminado de la base de datos.


> **Instrucciones**

**Cree una función** que tome un `DataFrame` y que le permita revisar si las diferentes variables que lo componen son coherentes y si es posible observar anomalías en estas.

**To-Do**
1. [ ] Imprimir un mensaje con las dimensiones del Dataframe.
2. [ ] Imprimir un mensaje con los nombres de las columnas que conforman el DataFrame.
3. [ ] Mostrar los primeros 5, los últimos 5 elementos del DataFrame y realice un muestreo de 5 elementos de forma aleatoria.
4. [ ] Mostrar una descripción rápida de las variables numéricas del DataFrame (utilice un método visto en clases).
5. [ ] Imprimir un mensaje con la cantidad de valores nulos (investigue el método `isna()` y `sum()` de los DataFrames).
6. [ ] Imprimir un mensaje que detalle la cantidad de valores unicos (investigue `nunique()`) .
7. [ ] Imprimir un mensaje con el total de filas que contengan nombres duplicados (investigue `duplicated()` y su parámetro `subset`).
8. [ ] Mostrar las filas que contengan nombres sean duplicados (investigue `duplicated()` y su parámetro `subset`).

Finalmente, responda las siguientes preguntas (comente esto en la celda ubicada debajo de **justificación**.):

9. [ ] ¿Qué anomalías es posible observar de los datos que componen el DataFrame? ¿Existen valores nulos o duplicados?


10. [ ] Su compañero ChatGPT le provee un código para que pueda entender los resultados de las pruebas desde un gráfico. ¿Qué puede interpretar de la distribución de las notas en el boxplot?

> **Nota:** Para imprmir utilice `print(...)`. Para mostrar `DataFrames`, utilice `display(...)` ya que este permite mostrar estos elementos como tablas.

> **Nota 2:** Utilice la documentación de [pandas](https://pandas.pydata.org/docs/index.html) como también StackOverflow para resolver cualquier problema o inconveniente.

> **Nota 3:** Si no ha trabajado con boxplot antes, el siguiente link puede serle de mucha ayuda: [boxplot](https://medium.com/dayem-siddiqui/understanding-and-interpreting-box-plots-d07aab9d1b6c)

**Respuesta:**

In [5]:
def exploratory_data_analysis(dataframe):
    print(f'1.- El dataset tiene {len(dataframe)} filas y {len(dataframe.columns)} columnas')
    print(' ')
    print(f'2.- El DataFrame esta compuesto por las siguientes columnas: {list(dataframe.columns)}')
    print(' ')
    print('3.- Ejemplos de filas del DataFrame:')
    print(' ')
    print('Primeras 5 filas:')
    print(' ')
    display(dataframe.head(5))
    print(' ')
    print('Últimas 5 filas:')
    print(' ')
    display(dataframe.tail(5))
    print(' ')
    print('Muestreo aleatorio de 5 filas:')
    print(' ')
    display(dataframe.sample(5, random_state=42))
    print(' ')
    print('4.- Descripción numérica del Dataframe:')
    print(' ')
    display(dataframe.describe())
    print(' ')
    print('5.- Cantidad de valores nulos por columna:')
    print(dataframe.isna().sum())
    print(' ')
    print('6.- Cantidad de valores únicos por columna:')
    print(dataframe.nunique())
    print(' ')
    print(f'7.- Presenta un total de {dataframe.duplicated(subset=["names"]).sum()} filas con nombres duplicados.')
    print(' ')
    print('8.- Las filas con nombres duplicados son:')
    duplicated_rows = dataframe[dataframe.duplicated(subset=['names'])]
    display(duplicated_rows)

In [6]:
# Código del amigo ChatGPT (listo para usar)
fig = px.box(
    df_grades.melt(
        value_vars=["math score", "reading score", "writing score"],
        var_name="subject",
        value_name="score",
    ),
    y="score",
    color="subject",
)
fig.show()

**Justificación:**

Primero, es necesario observar que el dataframe tiene 1400 columnas (es decir 1400 nombres), pero sólo hay 1153 nombres únicos. Por lo que hay 247 filas donde se tiene el mismo nombre. Es importante mencionar que no necesariamente esto significa que exista más de una vez la misma información, pues dos personas pueden tener el mismo nombre. Por lo que habrá que explorar si la otra información también es la misma. Por otra parte, hay presencia de valores nulos en la columnas score. Además, hay máximos en los scores que no corresponden a la escala.

En cuanto a los boxplot, lo primero que se nota es que hay máximos que se escapan de la escla de puntaje. Además, se pueden ver que los tres scores están bastante cercanos entre si. El promedio de reading score (76) es levemente mayor al de writing score (75), debido probablemente a los scores minimos de esta última.

In [7]:
exploratory_data_analysis(df_grades)

1.- El dataset tiene 1400 filas y 9 columnas
 
2.- El DataFrame esta compuesto por las siguientes columnas: ['names', 'gender', 'race/ethnicity', 'parental level of education', 'lunch', 'test preparation course', 'math score', 'reading score', 'writing score']
 
3.- Ejemplos de filas del DataFrame:
 
Primeras 5 filas:
 


names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
1   Madeline Fuller  female        group C            some high school   
2     Daniel Rester    male        group E          associate's degree   
3  Charles Linstrom    male        group A           bachelor's degree   
4    Geraldine Hale  female        group A          associate's degree   

      lunch test preparation course  math score  reading score  writing score  
0  standard                    none        37.0           46.0           46.0  
1  standard                    none       124.0          142.0          142.0  
2  standard                    none       156.0          196.0          193.0  
3  standard               completed        80.0           78.0           81.0  
4  standard                    none       131.0          166.0          166.0

 
Últimas 5 filas:
 


names  gender race/ethnicity parental level of education  \
1395  Alexander Dillon    male        group D          associate's degree   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1398      Joseph Mccoy    male        group D                some college   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
1395      standard                    none       193.0          142.0   
1396  free/reduced               completed        66.0           74.0   
1397      standard               completed        54.0           61.0   
1398  free/reduced                    none       193.0          196.0   
1399      standard               completed        68.0           80.0   

      writing score  
1395          193.0  
1396           81.0  
1397           62.0  
1398          193.0  
1399           76.0

 
Muestreo aleatorio de 5 filas:
 


names  gender race/ethnicity parental level of education  \
665    Bonnie Williams  female        group A            some high school   
624        Marcia York  female        group D           bachelor's degree   
115  Francisco Johnson    male        group B          associate's degree   
478      Edward Lowman    male        group E          associate's degree   
233     Thomas Simmons    male        group C           bachelor's degree   

            lunch test preparation course  math score  reading score  \
665      standard               completed       193.0          196.0   
624  free/reduced                    none        63.0           73.0   
115      standard               completed        82.0           84.0   
478  free/reduced               completed        91.0           73.0   
233      standard               completed       131.0          171.0   

     writing score  
665           97.0  
624           78.0  
115           78.0  
478           80.0  
233          196.0

 
4.- Descripción numérica del Dataframe:
 


math score  reading score  writing score
count  1360.000000    1360.000000    1353.000000
mean     89.538971      93.086029      91.852919
std      47.152422      46.385775      47.043570
min       0.000000      17.000000      10.000000
25%      59.000000      63.000000      62.000000
50%      72.000000      76.000000      75.000000
75%      97.250000     100.000000     100.000000
max     196.000000     196.000000     196.000000

 
5.- Cantidad de valores nulos por columna:
names                           0
gender                          0
race/ethnicity                  0
parental level of education     0
lunch                           0
test preparation course         0
math score                     40
reading score                  40
writing score                  47
dtype: int64
 
6.- Cantidad de valores únicos por columna:
names                          1153
gender                            2
race/ethnicity                    5
parental level of education       6
lunch                             2
test preparation course           2
math score                       96
reading score                    88
writing score                    92
dtype: int64
 
7.- Presenta un total de 247 filas con nombres duplicados.
 
8.- Las filas con nombres duplicados son:


names  gender race/ethnicity parental level of education  \
12         Rita Courtney  female        group B            some high school   
81    Elizabeth Quintero  female        group B                some college   
160    Michelle Gonzales  female        group B            some high school   
170        Carla Mendoza  female        group D          associate's degree   
176       Darlene Parker  female        group C          associate's degree   
...                  ...     ...            ...                         ...   
1385      James Crawford    male        group D             master's degree   
1390     Gloria Bradford  female        group B                 high school   
1393         Adam Romero    male        group C                 high school   
1395    Alexander Dillon    male        group D          associate's degree   
1398        Joseph Mccoy    male        group D                some college   

             lunch test preparation course  math score  reading score  \
12        standard                    none       193.0          193.0   
81        standard                    none        82.0           85.0   
160       standard               completed        60.0           70.0   
170   free/reduced                    none        43.0           60.0   
176       standard                    none        54.0           61.0   
...            ...                     ...         ...            ...   
1385      standard                    none       193.0          147.0   
1390      standard                    none       124.0           64.0   
1393      standard               completed       124.0          193.0   
1395      standard                    none       193.0          142.0   
1398  free/reduced                    none       193.0          196.0   

      writing score  
12            193.0  
81             87.0  
160            74.0  
170            58.0  
176            58.0  
...             ...  
1385          131.0  
1390          193.0  
1393           49.0  
1395          193.0  
1398          193.0  

[247 rows x 9 columns]

**Respuesta Esperada:**

In [8]:
exploratory_data_analysis(df_grades)

1.- El dataset tiene 1400 filas y 9 columnas
 
2.- El DataFrame esta compuesto por las siguientes columnas: ['names', 'gender', 'race/ethnicity', 'parental level of education', 'lunch', 'test preparation course', 'math score', 'reading score', 'writing score']
 
3.- Ejemplos de filas del DataFrame:
 
Primeras 5 filas:
 


names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
1   Madeline Fuller  female        group C            some high school   
2     Daniel Rester    male        group E          associate's degree   
3  Charles Linstrom    male        group A           bachelor's degree   
4    Geraldine Hale  female        group A          associate's degree   

      lunch test preparation course  math score  reading score  writing score  
0  standard                    none        37.0           46.0           46.0  
1  standard                    none       124.0          142.0          142.0  
2  standard                    none       156.0          196.0          193.0  
3  standard               completed        80.0           78.0           81.0  
4  standard                    none       131.0          166.0          166.0

 
Últimas 5 filas:
 


names  gender race/ethnicity parental level of education  \
1395  Alexander Dillon    male        group D          associate's degree   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1398      Joseph Mccoy    male        group D                some college   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
1395      standard                    none       193.0          142.0   
1396  free/reduced               completed        66.0           74.0   
1397      standard               completed        54.0           61.0   
1398  free/reduced                    none       193.0          196.0   
1399      standard               completed        68.0           80.0   

      writing score  
1395          193.0  
1396           81.0  
1397           62.0  
1398          193.0  
1399           76.0

 
Muestreo aleatorio de 5 filas:
 


names  gender race/ethnicity parental level of education  \
665    Bonnie Williams  female        group A            some high school   
624        Marcia York  female        group D           bachelor's degree   
115  Francisco Johnson    male        group B          associate's degree   
478      Edward Lowman    male        group E          associate's degree   
233     Thomas Simmons    male        group C           bachelor's degree   

            lunch test preparation course  math score  reading score  \
665      standard               completed       193.0          196.0   
624  free/reduced                    none        63.0           73.0   
115      standard               completed        82.0           84.0   
478  free/reduced               completed        91.0           73.0   
233      standard               completed       131.0          171.0   

     writing score  
665           97.0  
624           78.0  
115           78.0  
478           80.0  
233          196.0

 
4.- Descripción numérica del Dataframe:
 


math score  reading score  writing score
count  1360.000000    1360.000000    1353.000000
mean     89.538971      93.086029      91.852919
std      47.152422      46.385775      47.043570
min       0.000000      17.000000      10.000000
25%      59.000000      63.000000      62.000000
50%      72.000000      76.000000      75.000000
75%      97.250000     100.000000     100.000000
max     196.000000     196.000000     196.000000

 
5.- Cantidad de valores nulos por columna:
names                           0
gender                          0
race/ethnicity                  0
parental level of education     0
lunch                           0
test preparation course         0
math score                     40
reading score                  40
writing score                  47
dtype: int64
 
6.- Cantidad de valores únicos por columna:
names                          1153
gender                            2
race/ethnicity                    5
parental level of education       6
lunch                             2
test preparation course           2
math score                       96
reading score                    88
writing score                    92
dtype: int64
 
7.- Presenta un total de 247 filas con nombres duplicados.
 
8.- Las filas con nombres duplicados son:


names  gender race/ethnicity parental level of education  \
12         Rita Courtney  female        group B            some high school   
81    Elizabeth Quintero  female        group B                some college   
160    Michelle Gonzales  female        group B            some high school   
170        Carla Mendoza  female        group D          associate's degree   
176       Darlene Parker  female        group C          associate's degree   
...                  ...     ...            ...                         ...   
1385      James Crawford    male        group D             master's degree   
1390     Gloria Bradford  female        group B                 high school   
1393         Adam Romero    male        group C                 high school   
1395    Alexander Dillon    male        group D          associate's degree   
1398        Joseph Mccoy    male        group D                some college   

             lunch test preparation course  math score  reading score  \
12        standard                    none       193.0          193.0   
81        standard                    none        82.0           85.0   
160       standard               completed        60.0           70.0   
170   free/reduced                    none        43.0           60.0   
176       standard                    none        54.0           61.0   
...            ...                     ...         ...            ...   
1385      standard                    none       193.0          147.0   
1390      standard                    none       124.0           64.0   
1393      standard               completed       124.0          193.0   
1395      standard                    none       193.0          142.0   
1398  free/reduced                    none       193.0          196.0   

      writing score  
12            193.0  
81             87.0  
160            74.0  
170            58.0  
176            58.0  
...             ...  
1385          131.0  
1390          193.0  
1393           49.0  
1395          193.0  
1398          193.0  

[247 rows x 9 columns]

## 2 Limpieza de Datos [1.0 puntos]




<p align="center">
  <img src="https://blogdatlas.files.wordpress.com/2020/07/datlas_mx_blog_limpieza_bds_data-scrubbing-service.gif">
</p>

En base a lo realizado anteriormente, realice una limpieza de los dataset. Asuma los supuestos señalados en la pregunta anterior para limpiar el `DataFrame` y cree una función que le permita limpiar los datos. Habiendo limpiado el `DataFrame`, compruebe nuevamente los datos con la función creada en la sección 1.1.

**To-Do:**

- [ ] Crear función que permita limpiar el dataset.
- [ ] Comprobar que no existen anomalías en los datos a traves de la función de la sección 1.1.

**Hint**

Puede ser útil los métodos ```.drop_duplicates()``` y ```.dropna()``` como también utilizar queries con máscaras booleanas.


In [9]:
def clean_data(dataframe):
  dataframe = dataframe.drop_duplicates(subset=['names'], keep='first')
  dataframe = dataframe.dropna()
  dataframe = dataframe[(dataframe['math score'] <= 100) & (dataframe['reading score'] <= 100) & (dataframe['writing score'] <= 100)]
  dataframe = dataframe.reset_index(drop=True)
  return dataframe

In [10]:
df_grades = clean_data(df_grades)
df_grades

names  gender race/ethnicity parental level of education  \
0       Rita Courtney  female        group B            some high school   
1    Charles Linstrom    male        group A           bachelor's degree   
2         Brian Young    male        group C            some high school   
3      Howard Jimenez    male        group E            some high school   
4        Wayne Wilson    male        group B            some high school   
..                ...     ...            ...                         ...   
870     Richard Young    male        group D                 high school   
871     Wanda Russell  female        group B                 high school   
872    Marina Zeigler  female        group C           bachelor's degree   
873     Laurie Carter  female        group B            some high school   
874      Amanda Perez  female        group A                 high school   

            lunch test preparation course  math score  reading score  \
0        standard                    none        37.0           46.0   
1        standard               completed        80.0           78.0   
2        standard                    none        73.0           66.0   
3        standard               completed        81.0           75.0   
4        standard               completed        94.0           86.0   
..            ...                     ...         ...            ...   
870      standard                    none        69.0           75.0   
871  free/reduced               completed        23.0           44.0   
872  free/reduced               completed        66.0           74.0   
873      standard               completed        54.0           61.0   
874      standard               completed        68.0           80.0   

     writing score  
0             46.0  
1             81.0  
2             63.0  
3             76.0  
4             87.0  
..             ...  
870           71.0  
871           36.0  
872           81.0  
873           62.0  
874           76.0  

[875 rows x 9 columns]

**Respuesta Esperada:**

In [ ]:
df_grades = clean_data(df_grades)
df_grades

names  gender race/ethnicity parental level of education  \
0       Rita Courtney  female        group B            some high school   
1    Charles Linstrom    male        group A           bachelor's degree   
2         Brian Young    male        group C            some high school   
3      Howard Jimenez    male        group E            some high school   
4        Wayne Wilson    male        group B            some high school   
..                ...     ...            ...                         ...   
870     Richard Young    male        group D                 high school   
871     Wanda Russell  female        group B                 high school   
872    Marina Zeigler  female        group C           bachelor's degree   
873     Laurie Carter  female        group B            some high school   
874      Amanda Perez  female        group A                 high school   

            lunch test preparation course  math score  reading score  \
0        standard                    none        37.0           46.0   
1        standard               completed        80.0           78.0   
2        standard                    none        73.0           66.0   
3        standard               completed        81.0           75.0   
4        standard               completed        94.0           86.0   
..            ...                     ...         ...            ...   
870      standard                    none        69.0           75.0   
871  free/reduced               completed        23.0           44.0   
872  free/reduced               completed        66.0           74.0   
873      standard               completed        54.0           61.0   
874      standard               completed        68.0           80.0   

     writing score  
0             46.0  
1             81.0  
2             63.0  
3             76.0  
4             87.0  
..             ...  
870           71.0  
871           36.0  
872           81.0  
873           62.0  
874           76.0  

[875 rows x 9 columns]

> Utilice la función creada en el punto 1.1 para verificar sus resultados.

In [11]:
exploratory_data_analysis(df_grades)

1.- El dataset tiene 875 filas y 9 columnas
 
2.- El DataFrame esta compuesto por las siguientes columnas: ['names', 'gender', 'race/ethnicity', 'parental level of education', 'lunch', 'test preparation course', 'math score', 'reading score', 'writing score']
 
3.- Ejemplos de filas del DataFrame:
 
Primeras 5 filas:
 


names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
1  Charles Linstrom    male        group A           bachelor's degree   
2       Brian Young    male        group C            some high school   
3    Howard Jimenez    male        group E            some high school   
4      Wayne Wilson    male        group B            some high school   

      lunch test preparation course  math score  reading score  writing score  
0  standard                    none        37.0           46.0           46.0  
1  standard               completed        80.0           78.0           81.0  
2  standard                    none        73.0           66.0           63.0  
3  standard               completed        81.0           75.0           76.0  
4  standard               completed        94.0           86.0           87.0

 
Últimas 5 filas:
 


names  gender race/ethnicity parental level of education  \
870   Richard Young    male        group D                 high school   
871   Wanda Russell  female        group B                 high school   
872  Marina Zeigler  female        group C           bachelor's degree   
873   Laurie Carter  female        group B            some high school   
874    Amanda Perez  female        group A                 high school   

            lunch test preparation course  math score  reading score  \
870      standard                    none        69.0           75.0   
871  free/reduced               completed        23.0           44.0   
872  free/reduced               completed        66.0           74.0   
873      standard               completed        54.0           61.0   
874      standard               completed        68.0           80.0   

     writing score  
870           71.0  
871           36.0  
872           81.0  
873           62.0  
874           76.0

 
Muestreo aleatorio de 5 filas:
 


names  gender race/ethnicity parental level of education  \
797      Donna Duarte  female        group C                 high school   
319  Edward Hutcheson    male        group A             master's degree   
382    Louis Mckinzie    male        group C          associate's degree   
790     Allen Schwarz    male        group C            some high school   
422       Kara Greene  female        group D            some high school   

            lunch test preparation course  math score  reading score  \
797      standard                    none        65.0           69.0   
319  free/reduced                    none        73.0           74.0   
382      standard                    none        74.0           73.0   
790      standard               completed        63.0           60.0   
422      standard               completed        61.0           74.0   

     writing score  
797           67.0  
319           72.0  
382           67.0  
790           57.0  
422           72.0

 
4.- Descripción numérica del Dataframe:
 


math score  reading score  writing score
count  875.000000     875.000000     875.000000
mean    66.225143      69.195429      68.114286
std     15.076716      14.629270      15.174220
min      0.000000      17.000000      10.000000
25%     57.000000      59.000000      57.500000
50%     66.000000      70.000000      69.000000
75%     76.500000      79.000000      79.000000
max    100.000000     100.000000     100.000000

 
5.- Cantidad de valores nulos por columna:
names                          0
gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
dtype: int64
 
6.- Cantidad de valores únicos por columna:
names                          875
gender                           2
race/ethnicity                   5
parental level of education      6
lunch                            2
test preparation course          2
math score                      79
reading score                   72
writing score                   77
dtype: int64
 
7.- Presenta un total de 0 filas con nombres duplicados.
 
8.- Las filas con nombres duplicados son:


Empty DataFrame
Columns: [names, gender, race/ethnicity, parental level of education, lunch, test preparation course, math score, reading score, writing score]
Index: []

In [12]:
# Código del amigo ChatGPT (listo para usar)
fig = px.box(
    df_grades.melt(
        value_vars=["math score", "reading score", "writing score"],
        var_name="subject",
        value_name="score",
    ),
    y="score",
    color="subject",
)
fig.show()

**Justificación:**

Se puede observar que luego de hacer la limpieza, no hay filas con nombres duplicados ni tampoco existen valores nulos en la data. Además, el máximo de los scores está dentro de la escala.

## 3. Trabajando con Datos [2.0 puntos]

<p align="center">
  <img src="https://media3.giphy.com/media/YQitE4YNQNahy/giphy-downsized-large.gif">
</p>

### 3.1 Transformación de notas a "Nota Chilena" [0.3 Puntos]

En una de las reuniones con Don Caguayo este les comenta su intención de pasar todas sus notas a la "Escala Chilena". Para esto les facilita la siguiente ecuación:

$$Nota \, Chilena= \% \, de \, logro*0.06 + 1$$

Utilizando la ecuación entregada, cree una función que le permita pasar a la escala Chilena las notas del dataframe y entregue las notas redondeadas al segundo decimal.

**To-Do**

- [ ] Tranforme los valores a la escala chilena.

**Respuesta:**

In [13]:
def nota_chilena(dataframe_in, columna):
  dataframe_in[columna] = (dataframe_in[columna] * 0.06 + 1).round(2)
  return dataframe_in

In [14]:
nota_chilena(df_grades, 'math score')
nota_chilena(df_grades, 'reading score')
nota_chilena(df_grades, 'writing score')
df_grades

names  gender race/ethnicity parental level of education  \
0       Rita Courtney  female        group B            some high school   
1    Charles Linstrom    male        group A           bachelor's degree   
2         Brian Young    male        group C            some high school   
3      Howard Jimenez    male        group E            some high school   
4        Wayne Wilson    male        group B            some high school   
..                ...     ...            ...                         ...   
870     Richard Young    male        group D                 high school   
871     Wanda Russell  female        group B                 high school   
872    Marina Zeigler  female        group C           bachelor's degree   
873     Laurie Carter  female        group B            some high school   
874      Amanda Perez  female        group A                 high school   

            lunch test preparation course  math score  reading score  \
0        standard                    none        3.22           3.76   
1        standard               completed        5.80           5.68   
2        standard                    none        5.38           4.96   
3        standard               completed        5.86           5.50   
4        standard               completed        6.64           6.16   
..            ...                     ...         ...            ...   
870      standard                    none        5.14           5.50   
871  free/reduced               completed        2.38           3.64   
872  free/reduced               completed        4.96           5.44   
873      standard               completed        4.24           4.66   
874      standard               completed        5.08           5.80   

     writing score  
0             3.76  
1             5.86  
2             4.78  
3             5.56  
4             6.22  
..             ...  
870           5.26  
871           3.16  
872           5.86  
873           4.72  
874           5.56  

[875 rows x 9 columns]

**Respuesta esperada:**

In [15]:
df_grades

names  gender race/ethnicity parental level of education  \
0       Rita Courtney  female        group B            some high school   
1    Charles Linstrom    male        group A           bachelor's degree   
2         Brian Young    male        group C            some high school   
3      Howard Jimenez    male        group E            some high school   
4        Wayne Wilson    male        group B            some high school   
..                ...     ...            ...                         ...   
870     Richard Young    male        group D                 high school   
871     Wanda Russell  female        group B                 high school   
872    Marina Zeigler  female        group C           bachelor's degree   
873     Laurie Carter  female        group B            some high school   
874      Amanda Perez  female        group A                 high school   

            lunch test preparation course  math score  reading score  \
0        standard                    none        3.22           3.76   
1        standard               completed        5.80           5.68   
2        standard                    none        5.38           4.96   
3        standard               completed        5.86           5.50   
4        standard               completed        6.64           6.16   
..            ...                     ...         ...            ...   
870      standard                    none        5.14           5.50   
871  free/reduced               completed        2.38           3.64   
872  free/reduced               completed        4.96           5.44   
873      standard               completed        4.24           4.66   
874      standard               completed        5.08           5.80   

     writing score  
0             3.76  
1             5.86  
2             4.78  
3             5.56  
4             6.22  
..             ...  
870           5.26  
871           3.16  
872           5.86  
873           4.72  
874           5.56  

[875 rows x 9 columns]

### 3.2 Obtención de promedio de alumnos [0.3 Puntos]

Con todas las notas adjuntadas, la prestigiosa Universidad De la Cachaña necesita obtener el promedio general de cada uno de los alumnos. Para esto se le solicita que calcule el promedio con cada una de las asignaturas y el promedio se anexe como una nueva columna al final del `DataFrame` original. Cabe señalar que las notas deben ser redondeadas al segundo decimal.

**To-Do:**
- [ ] Crear función que nos permite obtener el promedio de cada uno de los alumnos y anexar esta información en el `DataFrame` original.


**Respuesta:**

In [16]:
def alumnos_promedio(dataframe_in):
  dataframe_in['GPA'] = dataframe_in[['math score', 'reading score', 'writing score']].mean(axis=1).round(2)
  return dataframe_in

In [17]:
df_grades = alumnos_promedio(df_grades)
df_grades.head()

names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
1  Charles Linstrom    male        group A           bachelor's degree   
2       Brian Young    male        group C            some high school   
3    Howard Jimenez    male        group E            some high school   
4      Wayne Wilson    male        group B            some high school   

      lunch test preparation course  math score  reading score  writing score  \
0  standard                    none        3.22           3.76           3.76   
1  standard               completed        5.80           5.68           5.86   
2  standard                    none        5.38           4.96           4.78   
3  standard               completed        5.86           5.50           5.56   
4  standard               completed        6.64           6.16           6.22   

    GPA  
0  3.58  
1  5.78  
2  5.04  
3  5.64  
4  6.34

**Respuesta Esperada:**

In [18]:
df_grades = alumnos_promedio(df_grades)
df_grades.head()

names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
1  Charles Linstrom    male        group A           bachelor's degree   
2       Brian Young    male        group C            some high school   
3    Howard Jimenez    male        group E            some high school   
4      Wayne Wilson    male        group B            some high school   

      lunch test preparation course  math score  reading score  writing score  \
0  standard                    none        3.22           3.76           3.76   
1  standard               completed        5.80           5.68           5.86   
2  standard                    none        5.38           4.96           4.78   
3  standard               completed        5.86           5.50           5.56   
4  standard               completed        6.64           6.16           6.22   

    GPA  
0  3.58  
1  5.78  
2  5.04  
3  5.64  
4  6.34

### 3.3 Reprobados y Destacados [0.9 Puntos, 0.3 cada DataFrame]

Contentos con su desempeño, la Universidad les solicita que obtengan los alumnos quienes han reprobado el año y los que han tenido desempeño de destacado. Para esto se considerará que un alumno ha reprobado el año cuando el promedio de notas es inferior a 4.0 en la Escala Chilena, y se considerará destacado si posee notas superiores a 5.8 en todas las asignaturas (o sea notas superiores a 5.8 en `math score`, `reading score` y `writing score`).

Cree una función que permita obtener tres Dataframes:
- El primero deberá poseer solamente los alumnos que han reprobado en la prestigiosa universidad.
- El segundo debe entregar un DataFrame similar al original, pero marcado con una "R" el promedio de notas de los alumnos reprobados.
- El tercero será un Data Frame con los alumnos destacados ordenados de mayor a menor.

**To-Do:**

- [ ] Obtener DataFrame con los alumnos que reprobaron el año.
- [ ] Obtener DataFrame similar al original, pero marcados con una "R" el atributo GPA de los alumnos reprobados.
- [ ] Obtener DataFrame con los alumnos destacados ordenados de mayor a menor.

**Respuesta:**

In [19]:
def reprobados_destacados(dataframe_in):
  reprobados = dataframe_in[dataframe_in['GPA'] < 4.0]

  df_reprobados = dataframe_in.copy()
  df_reprobados.loc[df_reprobados.index.isin(reprobados.index), 'GPA'] = 'R'

  destacados = dataframe_in[(dataframe_in['math score'] > 5.8) & (dataframe_in['reading score'] > 5.8) & (dataframe_in['writing score'] > 5.8)].sort_values(by=['GPA'], ascending=False)

  reprobados = reprobados.reset_index(drop=True)
  destacados = destacados.reset_index(drop=True)

  return reprobados, df_reprobados, destacados

In [20]:
dataframe1, dataframe2, dataframe3 = reprobados_destacados(df_grades)
dataframe1

names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
1       Mikki Amaya  female        group A                 high school   
2     Samuel Dougal    male        group C             master's degree   
3   Rodger Talavera    male        group B                 high school   
4    Morris Johnson    male        group A          associate's degree   
..              ...     ...            ...                         ...   
81   Charles Harian    male        group C            some high school   
82    Wendy Gunnell  female        group D            some high school   
83      Diana Costa  female        group C                 high school   
84      Amy Sieving  female        group C                 high school   
85    Wanda Russell  female        group B                 high school   

           lunch test preparation course  math score  reading score  \
0       standard                    none        3.22           3.76   
1   free/reduced               completed        3.04           3.88   
2   free/reduced               completed        3.76           3.52   
3   free/reduced                    none        3.16           2.74   
4   free/reduced               completed        3.40           4.30   
..           ...                     ...         ...            ...   
81  free/reduced               completed        3.70           4.12   
82  free/reduced                    none        2.62           3.04   
83  free/reduced                    none        3.46           3.76   
84      standard                    none        2.74           2.74   
85  free/reduced               completed        2.38           3.64   

    writing score   GPA  
0            3.76  3.58  
1            3.46  3.46  
2            3.76  3.68  
3            2.62  2.84  
4            4.18  3.96  
..            ...   ...  
81           3.94  3.92  
82           2.92  2.86  
83           3.58  3.60  
84           2.80  2.76  
85           3.16  3.06  

[86 rows x 10 columns]

In [21]:
dataframe2

names  gender race/ethnicity parental level of education  \
0       Rita Courtney  female        group B            some high school   
1    Charles Linstrom    male        group A           bachelor's degree   
2         Brian Young    male        group C            some high school   
3      Howard Jimenez    male        group E            some high school   
4        Wayne Wilson    male        group B            some high school   
..                ...     ...            ...                         ...   
870     Richard Young    male        group D                 high school   
871     Wanda Russell  female        group B                 high school   
872    Marina Zeigler  female        group C           bachelor's degree   
873     Laurie Carter  female        group B            some high school   
874      Amanda Perez  female        group A                 high school   

            lunch test preparation course  math score  reading score  \
0        standard                    none        3.22           3.76   
1        standard               completed        5.80           5.68   
2        standard                    none        5.38           4.96   
3        standard               completed        5.86           5.50   
4        standard               completed        6.64           6.16   
..            ...                     ...         ...            ...   
870      standard                    none        5.14           5.50   
871  free/reduced               completed        2.38           3.64   
872  free/reduced               completed        4.96           5.44   
873      standard               completed        4.24           4.66   
874      standard               completed        5.08           5.80   

     writing score   GPA  
0             3.76     R  
1             5.86  5.78  
2             4.78  5.04  
3             5.56  5.64  
4             6.22  6.34  
..             ...   ...  
870           5.26   5.3  
871           3.16     R  
872           5.86  5.42  
873           4.72  4.54  
874           5.56  5.48  

[875 rows x 10 columns]

In [22]:
dataframe3

names  gender race/ethnicity parental level of education  \
0       Richard Pryor    male        group E           bachelor's degree   
1       Sandra Pompey  female        group E          associate's degree   
2   Elizabeth Beasley  female        group E           bachelor's degree   
3           Emma Gray  female        group E           bachelor's degree   
4        Emma Brasher  female        group D            some high school   
..                ...     ...            ...                         ...   
89     Rolando Widmer    male        group C                 high school   
90  Lucille Stanphill  female        group C                 high school   
91       Steven Kelly    male        group D                some college   
92        Peter Perez    male        group D             master's degree   
93        Harry Berry    male        group B          associate's degree   

       lunch test preparation course  math score  reading score  \
0   standard               completed        7.00           7.00   
1   standard                    none        7.00           7.00   
2   standard                    none        7.00           7.00   
3   standard               completed        6.94           7.00   
4   standard               completed        6.82           7.00   
..       ...                     ...         ...            ...   
89  standard               completed        5.92           6.04   
90  standard                    none        5.86           6.04   
91  standard                    none        5.86           5.92   
92  standard                    none        5.86           5.86   
93  standard               completed        5.86           5.92   

    writing score   GPA  
0            7.00  7.00  
1            7.00  7.00  
2            7.00  7.00  
3            7.00  6.98  
4            7.00  6.94  
..            ...   ...  
89           5.92  5.96  
90           5.92  5.94  
91           6.04  5.94  
92           6.04  5.92  
93           5.92  5.90  

[94 rows x 10 columns]

**Respuesta:**

```
Escriba aqui su respuesta
```

### 3.4 Quintiles de Rendimiento [0.5 Puntos]

El rector, intrigado por saber si existe una relación entre el nivel educacional de los padres y las notas de sus hijos les plantea estudiar dicha relación. Para esto, les solicita que desde el `DataFrame` que posee a todos los alumnos (`DataFrame` obtenido tras la limpieza), generen una nueva columna que clasifique en quintiles de rendimiento para sus alumnos: `['Bajísimo', 'Bajo', 'Medio', 'Alto', 'Muy alto']`. (investigar `qcut()`)

Hecho esto, grafique a través de un scatter plot el nivel educacional de los padres v/s el promedio de los alumnos, utilice los cuantiles generados anteriormente para mostrar en que cuantil se encuentra cada uno de los datos.

Realizado el grafico, responda la siguiente pregunta: **¿Es posible visualizar alguna tendencia en el gráfico?.**



**Respuesta:**

In [23]:
# Código para quintiles


"""
NOTA: La variable del dataframe que contenga los quintiles debe llamarse
'GPA cuantiles' para que el gráfico funcione.
"""

def clasificar_quintiles(dataframe_in):
  dataframe_in['GPA cuantiles'] = pd.qcut(dataframe_in['GPA'], q=5, labels=['Bajísimo', 'Bajo', 'Medio', 'Alto', 'Muy alto'])
  return dataframe_in

In [26]:
df_grades = clasificar_quintiles(df_grades)

In [27]:
# Código de gráfico del amigo ChatGPT
fig = px.scatter(
    df_grades.sort_values(by="GPA"),
    x="GPA",
    y="parental level of education",
    hover_data=["gender"],
    color="GPA cuantiles",
    title="GPA por Nivel de educación de los padres",
    color_discrete_sequence=["tomato", "orange", "yellow", "greenyellow", "green"]
)

fig


Al visualizar los cuantiles más altos en todos los niveles, realmente no se nota ninguna tendencia. Pero si se observan los más bajos, hay una tendencia de que mientras mayor es el nivel de educación de los padres, más alto es el GPA.

## 4. Operaciones con tablas [2.0 puntos]

<p align="center">
  <img src="https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExYWx5dTBkMGhhOTJmeGQzbDV5M21rNTN1NWVwMm1qbHd3cjh3Nmk4NSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/13HBDT4QSTpveU/giphy.gif">
</p>



### 4.1 Carga de datos y concatenar dataset [0.2 Puntos]

Ya finalizado en análisis inicial, ud. y su equipo le entregaron a *Don Caguayo* (rector de la Universidad de la Cachaña) tanto los resultados del análisis como también la base de datos limpia y lista para ser almacenada. Dada la ingente cantidad de los datos, el equipo de TI de la universidad resolvió separar el dataset en dos bases de datos distintas (lo que según argumentan ellos, permitiría hacer agregaciones de forma más eficiente).

Gracias a la excelente labor de ud. y su equipo en el análisis previo, el rector le solicita continuar el trabajo con una nueva batería de análisis. Por este motivo, la sección de TI les entrega nuevamente los datos. Sin embargo, argumentan que dada una escazes de personal, solo le entregarán *dumps* (copias) de cada base de datos y su equipo deberá unir las bases de datos. Los datos se encuentran en los siguiente archivos `.json`: `students_grades_1.json` y `students_grades_2.json`.


Por ende, ud. y su equipo deciden que la primera tarea se centrará en cargar estos datos y unirlos.

**No se preocupe por la limpieza ni transformar el tipo de datos de las columnas, ni tampoco transformar a notas chilenas**, recuerde que anteriormente ya se encargo de este tema.


Hint: investigue el método `concat`

In [28]:
df_grades_1 =pd.read_json ('/content/drive/MyDrive/Lab programación cientifica/lab3/students_grades_1.json')
display(df_grades_1)
df_grades_2 =pd.read_json ('/content/drive/MyDrive/Lab programación cientifica/lab3/students_grades_2.json')
display(df_grades_2)

names  gender race/ethnicity parental level of education  \
0       Rita Courtney  female        group B            some high school   
1    Charles Linstrom    male        group A           bachelor's degree   
2         Brian Young    male        group C            some high school   
3      Howard Jimenez    male        group E            some high school   
4        Wayne Wilson    male        group B            some high school   
..                ...     ...            ...                         ...   
395    Vincent Tucker    male        group C                 high school   
396      Janet Housel  female        group B          associate's degree   
397   Gloria Bradford  female        group B                 high school   
398  Arthur Patterson    male        group A          associate's degree   
399     Ronald Wilson    male        group C            some high school   

            lunch test preparation course  math score  reading score  \
0        standard                    none        3.22           3.76   
1        standard               completed        5.80           5.68   
2        standard                    none        5.38           4.96   
3        standard               completed        5.86           5.50   
4        standard               completed        6.64           6.16   
..            ...                     ...         ...            ...   
395  free/reduced                    none        4.48           4.66   
396      standard                    none        5.38           5.56   
397      standard                    none        4.24           4.84   
398      standard                    none        4.24           4.18   
399      standard                    none        4.84           4.48   

     writing score  
0             3.76  
1             5.86  
2             4.78  
3             5.56  
4             6.22  
..             ...  
395           4.12  
396           5.80  
397           5.08  
398           3.82  
399           4.06  

[400 rows x 9 columns]

names  gender race/ethnicity parental level of education  \
0       Cheryl Kraus  female        group D            some high school   
1       Kristy Glass  female        group D          associate's degree   
2    Timothy Botello    male        group D             master's degree   
3       Ronald Coria    male        group A          associate's degree   
4     Jessica Martin  female        group B          associate's degree   
..               ...     ...            ...                         ...   
470    Richard Young    male        group D                 high school   
471    Wanda Russell  female        group B                 high school   
472   Marina Zeigler  female        group C           bachelor's degree   
473    Laurie Carter  female        group B            some high school   
474     Amanda Perez  female        group A                 high school   

            lunch test preparation course  math score  reading score  \
0        standard                    none        4.54           5.32   
1        standard                    none        5.56           5.44   
2        standard                    none        6.10           6.04   
3        standard               completed        6.82           6.52   
4    free/reduced                    none        4.18           5.20   
..            ...                     ...         ...            ...   
470      standard                    none        5.14           5.50   
471  free/reduced               completed        2.38           3.64   
472  free/reduced               completed        4.96           5.44   
473      standard               completed        4.24           4.66   
474      standard               completed        5.08           5.80   

     writing score  
0             5.80  
1             5.38  
2             6.34  
3             6.16  
4             5.20  
..             ...  
470           5.26  
471           3.16  
472           5.86  
473           4.72  
474           5.56  

[475 rows x 9 columns]

In [29]:
df_grades = pd.concat([df_grades_1, df_grades_2], ignore_index=True)
display(df_grades)

names  gender race/ethnicity parental level of education  \
0       Rita Courtney  female        group B            some high school   
1    Charles Linstrom    male        group A           bachelor's degree   
2         Brian Young    male        group C            some high school   
3      Howard Jimenez    male        group E            some high school   
4        Wayne Wilson    male        group B            some high school   
..                ...     ...            ...                         ...   
870     Richard Young    male        group D                 high school   
871     Wanda Russell  female        group B                 high school   
872    Marina Zeigler  female        group C           bachelor's degree   
873     Laurie Carter  female        group B            some high school   
874      Amanda Perez  female        group A                 high school   

            lunch test preparation course  math score  reading score  \
0        standard                    none        3.22           3.76   
1        standard               completed        5.80           5.68   
2        standard                    none        5.38           4.96   
3        standard               completed        5.86           5.50   
4        standard               completed        6.64           6.16   
..            ...                     ...         ...            ...   
870      standard                    none        5.14           5.50   
871  free/reduced               completed        2.38           3.64   
872  free/reduced               completed        4.96           5.44   
873      standard               completed        4.24           4.66   
874      standard               completed        5.08           5.80   

     writing score  
0             3.76  
1             5.86  
2             4.78  
3             5.56  
4             6.22  
..             ...  
870           5.26  
871           3.16  
872           5.86  
873           4.72  
874           5.56  

[875 rows x 9 columns]

### 4.2 Análisis por Nivel Educacional Etnia de los Padres [0.8 Puntos]

El rector, basado en su experiencia, cree fuertemente que el nivel educacional y la etnia de los padres influyen en las notas que obtienen sus hijos.
Como científicos de datos, ud. y su equipo creen que deben encontrar evidencia para confirmar o refutar la hipótesis del rector.

Para esto, deciden generar dos análisis: una tabla de resumen por una parte y gráficos de caja por otro.


**Tabla de Resumen [0.5 Puntos] **


Para generar la tabla de resumen:

- [ ] Calcular el promedio de las notas y guardarlo en una variable `GPA` (grade point average).
- [ ] Hacer una simplificación a través de un mapeo (investigar el método `map()`) de la variable `parental level of education`  según la siguiente conversión:

        some high school -> school
        some college -> school
        high school -> school
        bachelor's degree -> college
        associate's degree -> college
        master's degree -> postgraduate

  Los resultados de este mapeo deben ser guardados en la columna `simple parental level of education`.


- [ ] Agregar según 2 niveles: `race/ethnicity` y `simple parental level of education` para obtener el promedio de las notas.
- [ ] Agregar según 2 niveles: `race/ethnicity` y `simple parental level of education` para obtener un conteo de los alumnos en cada grupo y agregarlos como una nueva fila count.
- [ ] Obtener el porcentaje de alumnos con respecto al total. Los porcentajes deben ser strings que contienen la frecuencia de cada grupo con respecto al total y deben ser terminados en '%'.

Utilizar la tabla de resultados esperados como guía para desarrollar este punto.


**Análisis [0.3 Puntos]**

Finalmente, responda las siguientes preguntas sobre el gráfico de caja que viene implementado a continuación. En este se muestra el GPA con respecto al nivel educacional, donde la variable de color es la etnicidad.

- [ ]  ¿Hay alguna diferencia entre los grupos graficados tanto para el nivel educacional de los padres como también para la etnicidad?


In [30]:
df_grades['GPA'] = df_grades[['math score', 'reading score', 'writing score']].mean(axis=1)

mapeo = {
    'some high school': 'school',
    'some college': 'school',
    'high school': 'school',
    "bachelor's degree": 'college',
    "associate's degree": 'college',
    "master's degree": 'postgraduate'
}

df_grades['simple parental level of education'] = df_grades['parental level of education'].map(mapeo)

tabla_resumen = df_grades.groupby(['race/ethnicity', 'simple parental level of education']).agg({
    'math score': 'mean',
    'reading score': 'mean',
    'writing score': 'mean',
    'GPA': 'mean'
}).reset_index()

conteo_alumnos = df_grades.groupby(['race/ethnicity', 'simple parental level of education']).size().reset_index(name='count')

tabla_resumen = pd.merge(tabla_resumen, conteo_alumnos, on=['race/ethnicity', 'simple parental level of education'])

total_alumnos = df_grades.shape[0]
tabla_resumen['percentage'] = (tabla_resumen['count'] / total_alumnos * 100).round(2).astype(str) + ' %'

display(tabla_resumen)


race/ethnicity simple parental level of education  math score  \
0         group A                            college    4.735000   
1         group A                       postgraduate    4.690000   
2         group A                             school    4.568235   
3         group B                            college    5.070000   
4         group B                       postgraduate    4.912000   
5         group B                             school    4.689159   
6         group C                            college    5.021765   
7         group C                       postgraduate    4.924000   
8         group C                             school    4.764516   
9         group D                            college    5.108286   
10        group D                       postgraduate    5.218000   
11        group D                             school    5.018792   
12        group E                            college    5.542692   
13        group E                       postgraduate    5.540000   
14        group E                             school    5.396190   

    reading score  writing score       GPA  count percentage  
0        5.005000       4.890000  4.876667     24     2.74 %  
1        5.230000       5.350000  5.090000      2     0.23 %  
2        4.725882       4.558824  4.617647     51     5.83 %  
3        5.264444       5.192222  5.175556     54     6.17 %  
4        5.692000       5.548000  5.384000      5     0.57 %  
5        4.887664       4.764299  4.780374    107    12.23 %  
6        5.370588       5.354706  5.249020    102    11.66 %  
7        5.144000       5.104000  5.057333     15     1.71 %  
8        5.023871       4.919355  4.902581    155    17.71 %  
9        5.248000       5.248857  5.201714     70      8.0 %  
10       5.542000       5.734000  5.498000     20     2.29 %  
11       5.128725       5.110201  5.085906    149    17.03 %  
12       5.453846       5.450385  5.482308     52     5.94 %  
13       6.030000       5.890000  5.820000      6     0.69 %  
14       5.311429       5.159048  5.288889     63      7.2 %

In [31]:
fig = px.box(
    df_grades,
    x="simple parental level of education",
    y="GPA",
    color="race/ethnicity",
    hover_data=["GPA"],
)
fig.show()

En la figura se puede apreciar un patrón en el GPA de las distintas razas/etnias que se repite entre niveles educacionales. Por ejemplo, siempre el grupo E es el que tiene el GPA promedio más alto entre niveles educacionales, mientras que el grupo A siempre tiene el GPA promedio más bajo entre estos. Por lo que la de los padres raza/etnia si influye en el desempeño de sus hijos.

**Resultado Esperado de la tabla**

|    | race/ethnicity   | simple parental level of education   |   math score |   reading score |   writing score |   GPA |   count | percentage   |
|---:|:-----------------|:-------------------------------------|-------------:|----------------:|----------------:|------:|--------:|:-------------|
|  0 | group A          | college                              |         4.74 |            5    |            4.89 |  4.88 |      24 | 2.74 %       |
|  1 |           | postgraduate                         |         4.69 |            5.23 |            5.35 |  5.09 |       2 | 0.23 %       |
|  2 |           | school                               |         4.57 |            4.73 |            4.56 |  4.62 |      51 | 5.83 %       |
|  3 | group B          | college                              |         5.07 |            5.26 |            5.19 |  5.18 |      54 | 6.17 %       |
|  4 |           | postgraduate                         |         4.91 |            5.69 |            5.55 |  5.38 |       5 | 0.57 %       |
|  5 |           | school                               |         4.69 |            4.89 |            4.76 |  4.78 |     107 | 12.23 %      |
|  6 | group C          | college                              |         5.02 |            5.37 |            5.35 |  5.25 |     102 | 11.66 %      |
|  7 |           | postgraduate                         |         4.92 |            5.14 |            5.1  |  5.06 |      15 | 1.71 %       |
|  8 |           | school                               |         4.76 |            5.02 |            4.92 |  4.9  |     155 | 17.71 %      |
|  9 | group D          | college                              |         5.11 |            5.25 |            5.25 |  5.2  |      70 | 8.0 %        |
| 10 |           | postgraduate                         |         5.22 |            5.54 |            5.73 |  5.5  |      20 | 2.29 %       |
| 11 |           | school                               |         5.02 |            5.13 |            5.11 |  5.09 |     149 | 17.03 %      |
| 12 | group E          | college                              |         5.54 |            5.45 |            5.45 |  5.48 |      52 | 5.94 %       |
| 13 |          | postgraduate                         |         5.54 |            6.03 |            5.89 |  5.82 |       6 | 0.69 %       |
| 14 |          | school                               |         5.4  |            5.31 |            5.16 |  5.29 |      63 | 7.2 %        |

### 4.3 Joins, joins y más joins [1 punto]

Mientras le notificaba por videollamada los resultados de sus descubrimientos a Don Caguayo, un exaltado practicante del area de TI entra a la reunión y les informa que ha encontrado una nueva base de datos que cuenta con las notas de dos asignaturas (en escala chilena): historia y ciencias.
Para más remate, antes de huir, el practicante les cuenta que este dataframe lamentablemente contiene nuevamente los alumnos de los registros corruptos que ud. y su equipo filtraron en el análisis anterior.


El rector (evidentemente molesto por la situación) les ruega incluir estos datos (vaciados en el archivo other_grades.csv) al estudio original(`students_grades.csv`).


Para esto, carge el archivo `other_grades.csv` y busque la forma de unir ambos DataFrames, de tal manera que las columnas de `history score` y `science score` se anexen al final del DataFrame original. **NO LIMPIE LOS DATOS**, si no que explore los distintos tipos de merge para encontrar el mas situable para su situación (y así evitar buscar duplicados).

**To-Do**


- [ ] Cargar el `other_grades.csv`
- [ ] Unir `df_grades` con `other_grades.csv` usando outer join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando left join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando right join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando inner join y explique el resultado.
- [ ] Defina cuál join es el que utilizará para generar el nuevo DataFrame.

> **Hint**: Puede explicar los resultados del merge a través de la cantidad de filas resultantes y los valores que estas contienen.

In [32]:
df_other_grades =pd.read_csv('/content/drive/MyDrive/Lab programación cientifica/lab3/other_grades.csv')
display(df_other_grades)
df_grades =pd.read_json ('/content/drive/MyDrive/Lab programación cientifica/lab3/students_grades.json')
numeric_cols = ['math score', 'reading score', 'writing score']
df_grades[numeric_cols] = df_grades[numeric_cols].apply(pd.to_numeric, errors='coerce')
display(df_grades)

names  science score  history score
0           Pam Otoole             57            4.0
1          Diane Olson             48            4.8
2       Sarah Flannery             72            6.1
3     Stanley Chappell             47            3.7
4        Arnold Dodson             60            5.2
..                 ...            ...            ...
995    Melanie Acevedo             79            6.3
996   Christopher Wall             62            4.0
997  Margaret Spannaus             47            4.0
998     Vickie Sellers             47            5.7
999      Kari Mitchell             53            6.2

[1000 rows x 3 columns]

names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
1      Madeline Fuller  female        group C            some high school   
2        Daniel Rester    male        group E          associate's degree   
3     Charles Linstrom    male        group A           bachelor's degree   
4       Geraldine Hale  female        group A          associate's degree   
...                ...     ...            ...                         ...   
1395  Alexander Dillon    male        group D          associate's degree   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1398      Joseph Mccoy    male        group D                some college   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
0         standard                    none        37.0           46.0   
1         standard                    none       124.0          142.0   
2         standard                    none       156.0          196.0   
3         standard               completed        80.0           78.0   
4         standard                    none       131.0          166.0   
...            ...                     ...         ...            ...   
1395      standard                    none       193.0          142.0   
1396  free/reduced               completed        66.0           74.0   
1397      standard               completed        54.0           61.0   
1398  free/reduced                    none       193.0          196.0   
1399      standard               completed        68.0           80.0   

      writing score  
0              46.0  
1             142.0  
2             193.0  
3              81.0  
4             166.0  
...             ...  
1395          193.0  
1396           81.0  
1397           62.0  
1398          193.0  
1399           76.0  

[1400 rows x 9 columns]

In [33]:
outer_merge = pd.merge(df_grades, df_other_grades, how='outer', on='names')
display(outer_merge)

names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
1        Rita Courtney  female        group B            some high school   
2      Madeline Fuller  female        group C            some high school   
3        Daniel Rester    male        group E          associate's degree   
4     Charles Linstrom    male        group A           bachelor's degree   
...                ...     ...            ...                         ...   
1395     Richard Young    male        group D                 high school   
1396     Wanda Russell  female        group B                 high school   
1397    Marina Zeigler  female        group C           bachelor's degree   
1398     Laurie Carter  female        group B            some high school   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
0         standard                    none        37.0           46.0   
1         standard                    none       193.0          193.0   
2         standard                    none       124.0          142.0   
3         standard                    none       156.0          196.0   
4         standard               completed        80.0           78.0   
...            ...                     ...         ...            ...   
1395      standard                    none        69.0           75.0   
1396  free/reduced               completed        23.0           44.0   
1397  free/reduced               completed        66.0           74.0   
1398      standard               completed        54.0           61.0   
1399      standard               completed        68.0           80.0   

      writing score  science score  history score  
0              46.0           25.0            3.8  
1             193.0           25.0            3.8  
2             142.0            NaN            NaN  
3             193.0            NaN            NaN  
4              81.0           56.0            5.2  
...             ...            ...            ...  
1395           71.0           62.0            4.2  
1396           36.0           16.0            3.6  
1397           81.0           59.0            4.1  
1398           62.0           43.0            4.3  
1399           76.0           61.0            5.3  

[1400 rows x 11 columns]

El outer merge combina todos los elementos posibles y conserva todo el resto en filas independientes.  Incluye todas las filas de ambos DataFrames, completando con NaNs donde no haya coincidencia entre las columnas 'names'.

In [34]:
left_merge = pd.merge(df_grades, df_other_grades, how='left', on='names')
display(left_merge)

names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
1      Madeline Fuller  female        group C            some high school   
2        Daniel Rester    male        group E          associate's degree   
3     Charles Linstrom    male        group A           bachelor's degree   
4       Geraldine Hale  female        group A          associate's degree   
...                ...     ...            ...                         ...   
1395  Alexander Dillon    male        group D          associate's degree   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1398      Joseph Mccoy    male        group D                some college   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
0         standard                    none        37.0           46.0   
1         standard                    none       124.0          142.0   
2         standard                    none       156.0          196.0   
3         standard               completed        80.0           78.0   
4         standard                    none       131.0          166.0   
...            ...                     ...         ...            ...   
1395      standard                    none       193.0          142.0   
1396  free/reduced               completed        66.0           74.0   
1397      standard               completed        54.0           61.0   
1398  free/reduced                    none       193.0          196.0   
1399      standard               completed        68.0           80.0   

      writing score  science score  history score  
0              46.0           25.0            3.8  
1             142.0            NaN            NaN  
2             193.0            NaN            NaN  
3              81.0           56.0            5.2  
4             166.0            NaN            NaN  
...             ...            ...            ...  
1395          193.0           80.0            5.0  
1396           81.0           59.0            4.1  
1397           62.0           43.0            4.3  
1398          193.0           55.0            3.4  
1399           76.0           61.0            5.3  

[1400 rows x 11 columns]

El left merge conserva solo los elementos que se hayan combinado correctamente provenientes dataset `left`, en este caso `df_grades`. Incluye todas las filas de `df_grades`, completando con NaNs donde no haya coincidencia entre las columnas 'name' en `other_grades.csv`. Este tipo de unión mantiene todas las filas del DataFrame original `df_grades`, independientemente de si hay coincidencias en el DataFrame `other_grades.csv`.

In [36]:
right_merge = pd.merge(df_grades, df_other_grades, how='right', on='names')
display(right_merge)

names  gender race/ethnicity parental level of education  \
0            Pam Otoole  female        group B           bachelor's degree   
1           Diane Olson  female        group C                some college   
2        Sarah Flannery  female        group B             master's degree   
3      Stanley Chappell    male        group A          associate's degree   
4         Arnold Dodson    male        group C                some college   
...                 ...     ...            ...                         ...   
1242   Christopher Wall    male        group C                 high school   
1243  Margaret Spannaus  female        group C                 high school   
1244     Vickie Sellers  female        group D                some college   
1245      Kari Mitchell  female        group D                some college   
1246      Kari Mitchell  female        group D                some college   

             lunch test preparation course  math score  reading score  \
0         standard                    none        72.0           72.0   
1         standard               completed        69.0           90.0   
2         standard                    none        90.0           95.0   
3     free/reduced                    none        47.0           57.0   
4         standard                    none        76.0           78.0   
...            ...                     ...         ...            ...   
1242  free/reduced                    none        62.0           55.0   
1243  free/reduced               completed        59.0           71.0   
1244      standard               completed        68.0           78.0   
1245  free/reduced                    none        77.0           86.0   
1246  free/reduced                    none       193.0          193.0   

      writing score  science score  history score  
0              74.0             57            4.0  
1              88.0             48            4.8  
2              93.0             72            6.1  
3              44.0             47            3.7  
4              75.0             60            5.2  
...             ...            ...            ...  
1242           55.0             62            4.0  
1243           65.0             47            4.0  
1244           77.0             47            5.7  
1245           86.0             53            6.2  
1246          193.0             53            6.2  

[1247 rows x 11 columns]

El right merge conserva solo los elementos que se hayan combinado correctamente provenientes dataset `right`, en este caso `df_other_grades`. Este tipo de unión mantiene todas las filas del DataFrame other_grades.csv, independientemente de si hay coincidencias en el DataFrame original df_grades.

In [37]:
inner_merge = pd.merge(df_grades, df_other_grades, how='inner', on='names')
display(inner_merge)

names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
1        Rita Courtney  female        group B            some high school   
2     Charles Linstrom    male        group A           bachelor's degree   
3          Brian Young    male        group C            some high school   
4          Brian Young    male        group C            some high school   
...                ...     ...            ...                         ...   
1242     Richard Young    male        group D                 high school   
1243     Wanda Russell  female        group B                 high school   
1244    Marina Zeigler  female        group C           bachelor's degree   
1245     Laurie Carter  female        group B            some high school   
1246      Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
0         standard                    none        37.0           46.0   
1         standard                    none       193.0          193.0   
2         standard               completed        80.0           78.0   
3         standard                    none        73.0           66.0   
4         standard                    none         NaN            NaN   
...            ...                     ...         ...            ...   
1242      standard                    none        69.0           75.0   
1243  free/reduced               completed        23.0           44.0   
1244  free/reduced               completed        66.0           74.0   
1245      standard               completed        54.0           61.0   
1246      standard               completed        68.0           80.0   

      writing score  science score  history score  
0              46.0             25            3.8  
1             193.0             25            3.8  
2              81.0             56            5.2  
3              63.0             58            3.8  
4               NaN             58            3.8  
...             ...            ...            ...  
1242           71.0             62            4.2  
1243           36.0             16            3.6  
1244           81.0             59            4.1  
1245           62.0             43            4.3  
1246           76.0             61            5.3  

[1247 rows x 11 columns]

El inner join combina los elementos que se encuentren en ambas tablas. Descarta todo el resto. Este tipo de unión resulta en un DataFrame con menos filas que el DataFrame original `df_grades`, ya que solo incluye los registros que están presentes en ambas tablas.

Basandonos en los resultados y la necesidad del rector de incluir los nuevos datos al dataset ya trabajado, sin tener que limpiarlos nuevamente, se deberia ocupar un innerjoin, de esta forma se obtendrian los alumnos con las notas que se encuentran en ambas tablas, esto indirectamente aplicaria los diferentes filtros y limpieza al dataset agregado, por otra parte si se debe unir al dataset original que no se le ha aplicado ningun cambio, es mejor hacer un outerjoin para conservar la mayor cantidad de datos ya que junta los datos de ambas tablas llenando con nan donde no coinciden y posteriormente hacer la limpieza correspondiente.

# Conclusión


Eso ha sido todo para el lab de hoy, recuerden que el laboratorio tiene un plazo de entrega de una semana y que **los días de atraso no se pueden utilizar para entregas de lab** solo para tareas. Cualquier duda del laboratorio, no duden en contactarnos por mail o U-cursos.

<p align="center">
  <img src="https://media1.tenor.com/images/2f7df460196625edc45b8380d158e5ee/tenor.gif?itemid=16669403">
</p>